### PR TITLE
Modified lib/customer.rb, lib/order.rb, test/order_test.rb

### PR DESCRIPTION
**Pull request by cescarez**
# Assignment Submission: Grocery Store

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Question | Response
--- | ---
What is accomplished with `raise ArgumentError`? | An error is thrown and the program halts execution.
Why do you think we made the `.all` & `.find` class methods?  Why not instance methods? | `.all` and `.find` would not make sense to call within a Customer or Object instance. `.all` and `.find` are class methods instead since they are methods that do not have anything to do with individual instances. If these were instance methods, the method outputs would be the same regardless of the instance it was called on.
Think about the relation between `Order` and `Customer`. Is this relation one-to-one, one-to-many, or something else? How does that compare to the Solar System project? | The relation between `Order` and `Customer` is many-to-one since each `Customer` could have zero to many `Order`s, but each `Order` can only be associated with a single `Customer`. The Solar System project was also many to one, when considering the relation between `Planet` and `SolarSystem` since each `Planet` could only be in a single `SolarSystem` but a `SolarSystem` could have zero to many `Planet`s.
How is the relation between `Order` and `Customer` tracked in the CSV file? How is it tracked in your program? Why might these be different? | data/order.csv had the customer ID recorded in each order. In my program the relation between the classes is present in the ability for methods in Order to call Customer methods and instance variables. The `Customer` class is not "aware" of the `Order` class, or any instances.
Did the presence of automated tests change the way you thought about the problem? How? | Yes, automated tests allowed me to ensure that edge cases were considered and accounted for while writing my non-test source code.

### CODE REVIEW

File: lib/order.rb [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  attr_reader :id, :products, :fulfillment_status, :customer
+
+  def initialize(id_num, products, customer_object,fulfillment_status = VALID_STATUSES[0])
+    @id = id_num
+    @customer = customer_object
+    @products = (products && !products.empty?) ? products : {}
+    @fulfillment_status = check_fulfillment_status(fulfillment_status)
+  end
+
+  def check_fulfillment_status(current_status)
```
Comment (jmaddox19) [inline lib/order.rb:24]: Love the input validation here and even more so that it has its own little method!

File: lib/order.rb [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    end
+
+    return orders.sort_by { |order| order.id }
+  end
+
+  def self.find(id)
+    return all.bsearch { |order| id.to_i <=> order.id }
+  end
+
+  def self.find_by_customer(customer_id)
```
Comment (jmaddox19) [inline lib/order.rb:68]: Fun!

File: lib/order.rb [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    return all.filter { |order| order.customer.id == customer_id.to_i }
+  end
+
+  private
+
+  def self.get_all_products(external_file_data)
```
Comment (jmaddox19) [inline lib/order.rb:74]: It helps so much with readability to pull this out into a helper method. 
On top of that, you made it private which is perfect! Because it's only used within this class. This further helps with readability of the code.

File: lib/order.rb [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    all_products = get_all_products(external_orders_file)
+
+    orders = external_orders_file.each_with_index.map do |order, i|
+      Order.new(order[0].to_i, all_products[i], Customer.find(order[2].to_i), order[3].to_sym)
+    end
+
+    return orders.sort_by { |order| order.id }
+  end
+
+  def self.find(id)
```
Comment (jmaddox19) [inline lib/order.rb:64]: Fun little tidbit, now that we know about abstract classes, we can look back at this and observe that this would be a great candidate for a method in an abstract class, because it looks the same in Customer and Order :D

**Comment by jmaddox19** [comment]
# Grocery Store





## Major Learning Goals/Code Review



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
The `Customer` and `Order` classes have appropriate `attr_` helpers for instance variables | ✔️
The `CSV` library is used in `.all` methods, and not in `.find` methods | ✔️
Appropriately parses the product data from CSV file in `Order.all` | ✔️
`Order.all` calls `Customer.find` to set up the composition relation | ✔️
Practices git with at least 3 small commits and meaningful commit messages | ✔️ So many commits with details messages!

## Functional Requirements

| Functional Requirement | yes/no |
| --- | --- |
All of the wave 1 tests in `Customer` pass | ✔️
All of the wave 1 tests in `Order` pass | ✔️
All of the wave 2 tests in `Customer` pass | ✔️
All of the wave 2 tests in `Order` pass | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 4+ in Code Review && 3+ in Functional Requirements | ✔️
| Yellow (Approaches Standards) | 2+ in Code Review && 2+ in Functional Requirements | 
| Red (Not at Standard) | 0,1 in Code Review or 0,1 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging | 


The code looks AMAZING! This project clearly demonstrates that you learned all the week's learning goals! There are classes with appropriate relationships setup between them. Class methods are used appropriately and call other class methods when appropriate.

The code is very readable and concise! The input is validated thoroughly in a way that doesn't clutter the code. Helper methods are created everywhere appropriate. Great work! :)


## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅
